src/software/amazon/ion/Timestamp.java

```diff
@@ -700,12 +728,26 @@ private Timestamp(BigDecimal millis, Integer localOffset)
         }
         else {
             BigDecimal secs = millis.movePointLeft(3);
-            BigDecimal secsDown = secs.setScale(0, RoundingMode.FLOOR);
+            BigDecimal secsDown = fastZeroScaleRoundDown(secs);
             this._fraction = secs.subtract(secsDown);
```

Comment:
The use of `_fraction` in this class is very messy and error-prone, and entirely vestigial from times preceding the OSS release of this code.  It's long been my goal to remove it and turn `_second` into a `BigDecimal`.  Doing so would eliminate some of the challenges here.